Write up documentation on posthog-dotnet (#10589)
PostHog/posthog.com: contents/docs/libraries/dotnet/index.mdx

---
title: .NET
sidebarTitle: .NET
sidebar: Docs
showTitle: true
github: 'https://github.com/PostHog/posthog-dotnet'
icon: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/posthog.com/contents/images/docs/integrate/dotnet.svg
features:
  eventCapture: true
  userIdentification: true
  featureFlags: true
  groupAnalytics: true
  surveys: false
  llmObservability: false
  errorTracking: false
---

> <blockquote class="warning-note">
>     ⚠️  Warning - This is in beta and may break.
> </blockquote>

Reviewer: ```suggestion
> ⚠️ **Warning:** This is in beta and may break.
```

This is an optional library you can install if you're working with .NET Core. It uses an internal queue to make calls fast and non-blocking. It also batches requests and flushes asynchronously, making it perfect to use in any part of your web app or other server side application that needs performance.

## Installation

import DotNetInstall from '../../integrate/_snippets/install-dotnet.mdx'

<DotNetInstall />

> 

Reviewer: ```suggestion
```

    personPropertiesToSetOnce: new() { ["initial_url"] = "/blog" }
);
```

For more details on the difference between `$set` and `$set_once`, see our [person properties docs](/docs/data/user-properties#what-is-the-difference-between-set-and-set_once).

To capture [anonymous events](/docs/data/anonymous-vs-identified-events) without person profiles, set the event's `$process_person_profile` property to `false`:

```csharp
posthog.Capture(
    "distinct_id",
    "event_name",
    properties: new() {
        ["$process_person_profile"] = false
    }
)
```

## Alias

Sometimes, you want to assign multiple distinct IDs to a single user. This is helpful when your primary distinct ID is inaccessible. For example, if a distinct ID used on the frontend is not available in your backend. 

In this case, you can use `alias` to assign another distinct ID to the same user.

```csharp
await posthog.AliasAsync('current_distinct_id', 'new_distinct_id');
```

We strongly recommend reading our docs on [alias](/docs/product-analytics/identify#alias-assigning-multiple-distinct-ids-to-the-same-user) to best understand how to correctly use this method.

## Group analytics

> Group analytics allows you to associate an event with a group (e.g. teams, organizations, etc.). Read the [Group Analytics](/docs/user-guides/group-analytics) guide for more information.

Reviewer: ```suggestion
Group analytics allows you to associate an event with a group (e.g. teams, organizations, etc.). Read the [group analytics](/docs/product-analytics/group-analytics) guide for more information.
```

> **Note:** This is a paid feature and is not available on the open-source or free cloud plan. Learn more on our [pricing page](/pricing).

> -   Capture an event and associate it with a group

Reviewer: ```suggestion
To capture an event and associate it with a group, add the `groups` argument to your `Capture` call:
```

```csharp
postHogClient.Capture(
    "user_distinct_id", 
    "some_event", 
    groups: [new Group("company", "company_id_in_your_db")]);
```

> -   Update properties on a group

Reviewer: ```suggestion
Update properties on a group, use the `GroupIdentifyAsync` method:
```

```csharp
await postHogClient.GroupIdentifyAsync(
    type: "company",
    key: "company_id_in_your_db",
    name: "Awesome Inc.",
    properties: new()
    {
        ["employees"] = 11
    }
);
```

The `name` is a special property which is used in the PostHog UI for the name of the group. If you don't specify a `name` property, the group ID will be used instead.

## Feature flags

import FeatureFlagsLibsIntro from "../_snippets/feature-flags-libs-intro.mdx"

<FeatureFlagsLibsIntro />

import DotNetFeatureFlagsCode from '../../integrate/feature-flags-code/_snippets/feature-flags-code-dotnet.mdx'

<DotNetFeatureFlagsCode />

### Local evaluation

import LocalEvaluationIntro from "../../feature-flags/snippets/local-evaluation-intro.mdx"

<LocalEvaluationIntro />

For details on how to implement local evaluation, see our [local evaluation guide](/docs/feature-flags/local-evaluation).

## Experiments (A/B tests)

> Since [experiments](/docs/experiments/manual) use feature flags, the code for running an experiment is very similar to the feature flags code:

Reviewer: ```suggestion
Since [experiments](/docs/experiments/installation) use feature flags, the code for running an experiment is very similar to the feature flags code:
```

```csharp
if (await posthog.GetFeatureFlagAsync(
        "experiment-feature-flag-key",
        "user_distinct_id")
    is { VariantKey: "variant-name" })
{
    // Do something
}
```

It's also possible to [run experiments without using feature flags](/docs/experiments/running-experiments-without-feature-flags). 

> 

Reviewer: ```suggestion
```

## GeoIP properties

> The posthog-dotnet library disregards the server IP, does not add the GeoIP properties, and does not use the values for feature flag evaluations.

Reviewer: ```suggestion
The `posthog-dotnet` library disregards the server IP. This means it does not add the GeoIP properties and does not use the values for feature flag evaluations.
```

## Serverless environments (Azure Functions/Render/Lambda/...)

> By default, the library buffers events before sending them to the `/batch` endpoint, for better performance. This can
> lead to lost events in serverless environments, if the .NET process is terminated by the platform before the buffer
> is fully flushed. To avoid this, you can:
> 
> - ensure that `await posthog.FlushAsync()` is called after processing every request, by adding a middleware to your server: this allows `posthog.Capture()` to remain asynchronous for better performance.

Reviewer: ```suggestion
By default, the library buffers events before sending them to the `/batch` endpoint for better performance. This can lead to lost events in serverless environments if the .NET process is terminated by the platform before the buffer is fully flushed. 

To avoid this, ensure that `await posthog.FlushAsync()` is called after processing every request, by adding a middleware to your server. This allows `posthog.Capture()` to remain asynchronous for better performance.
```
